Add support for 64-bit numbers and size_t to wxConfig

vadz opened the pull request:
See [#19091](https://trac.wxwidgets.org/ticket/19091) but beyond this lack of support for 64-bit numbers is a bit ridiculuous in 2021.

vadz committed:
Get rid of CppUnit boilerplate in wxConfig unit tests

No real changes.

vadz committed:
Use smart pointer instead of manual delete in wxConfig tests

No real changes.

vadz committed:
Remove wxT() from wxConfig unit tests

Just remove the visual noise, no real changes.

vadz committed:
Use template function to avoid code duplication in wxRegConfig

Don't repeat exactly the same code for 3 different types, just use a
template function instead.

Change the existing TryGetValue() helpers to take all output parameters
consistently by pointer, rather than taking only long by pointer and the
other by reference. This allows to isolate the inconsistency of the type
of wxRegKey::QueryValue() parameter in a single place here, hiding it
from the rest of wxRegConfig code.

Add an extra unit test for writing/reading longs to/from wxRegConfig.

vadz committed:
Add 64-bit numbers support to wxRegKey

This uses the native registry support for 64-bit values available since
Windows XP.

Note that SetValue64() can't be just called SetValue() as this would
introduce ambiguities when writing an int or an enum to the registry.
This could be avoided by providing overloads for all the arithmetic
types, but it's arguably better to be more clear about what exactly is
being written to the registry in this low-level class and using a
different name is definitely simpler.

vadz committed:
Add 64-bit integers support to wxConfig

Serialize them to strings in wxFileConfig, just as we always did for
long, but use wxRegKey support for storing them directly to the registry
in wxRegConfig.

vadz committed:
Add support for size_t to wxConfig

Allow writing and reading size_t values directly too, there were not
previously accepted unlike all the other arithmetic types.

Closes #19091.

PBfordev commented on interface/wx/config.h:
```diff
@@ -549,6 +549,51 @@ class wxConfigBase : public wxObject
     bool Read(const wxString& key, long* l,
               long defaultVal) const;
+    /**
+        Reads a 64-bit long long value, returning @true if the value was found.
```
Perhaps everywhere where there is something like "value was/was not found", "value" should be replaced with "key"? 

I think "key" is already used like this in wxConfigBase docs and using "value" can bit a bit confusing when we also have an argument named "value" as well.

PBfordev replied:
I see, I just looked at the first couple of getters in the docs, they used "key".  But they were for `wxString`s, the ones for numbers do you use "value".

PBfordev commented on interface/wx/config.h:
```diff
@@ -549,6 +549,51 @@ class wxConfigBase : public wxObject
+        Not supported by wxPerl.
+        @endWxPerlOnly
+    */
+    bool Read(const wxString& key, wxLongLong_t* ll,
```
The function declaration seems to be incomplete, should probably look like
```suggestion
    bool Read(const wxString& key, wxLongLong_t* ll, wxLongLong_t defaultVal) const;
```

vadz replied:
Oops, thanks for noticing!

PBfordev commented on src/common/config.cpp:
```diff
@@ -174,6 +177,40 @@ bool wxConfigBase::Read(const wxString& key, int *pi, int defVal) const
 }
 
+// size_t is stored either as long or long long (Win64)
+#if SIZEOF_SIZE_T == SIZEOF_LONG
```
Sorry for nitpicking, but as `size_t` is unsigned, should we not used `unsigned long` and `wxULongLong` or `wxUint64` here?

But perhaps the compilers will not complain...

vadz replied:
Handling of unsigned types in `wxConfig` is suspicious generally speaking, but I didn't want to change it now, so I've just done the strict minimum needed to allow using `size_t` with it at all.

vadz commented:
Note: clang tests failure is unrelated.

vadz committed:
Compile wxConfig unit tests as part of non-GUI test too

Almost none of the tests require GUI, so make it possible to run them
without it, but still include the GUI tests when using GUI.
